hoomd/mpcd/CollisionMethod.cc

```diff
@@ -406,6 +432,158 @@ void mpcd::CollisionMethod::transferRigidBodyMomenta(uint64_t timestep)
+        const double3 cell_energy = h_cell_energy.data[bin_idx];
+        const unsigned int np = __double_as_int(cell_energy.z);
+
+        Scalar kinetic_energy
+            = cell_energy.x
+              + 0.5 * mass
+                    * (vel_mass.x * vel_mass.x + vel_mass.y * vel_mass.y + vel_mass.z * vel_mass.z);
+        Scalar temperature = 2 * kinetic_energy / (3 * (np - 1 + 1));
+
+        // thermostat relative velocities
+        // get the global index from the bin numbers found
+        const int3 global_cell = m_cl->getGlobalCell(bin);
+        const unsigned int global_idx = global_ci(global_cell.x, global_cell.y, global_cell.z);
+        // get the rng
+        uint16_t seed = m_sysdef->getSeed();
+        hoomd::RandomGenerator rng(
+            hoomd::Seed(hoomd::RNGIdentifier::CollisionMethod, timestep, seed),
+            hoomd::Counter(global_idx));
+
+        // the total number of degrees of freedom in the cell divided by 2
+        const double alpha = m_sysdef->getNDimensions() * (np - 1 + 1) / (double)2.;
+
+        // rescale velocity if there are other particles in the cell
+        if (alpha > 0)
+            {
+            // draw a random kinetic energy for the cell at the set temperature
+            hoomd::GammaDistribution<double> gamma_gen(alpha, 1.0);
+            const double rand_ke = gamma_gen(rng);
+
+            // generate the scale factor from the current temperature
+            // (don't use the kinetic energy of this cell, since this
+            // is total not relative to COM)
+            const double cur_ke = alpha * temperature;
+            double factor = (cur_ke > 0.) ? fast::sqrt(rand_ke / cur_ke) : 1.;
+
+            // rescale with the temperature of the cell
+            Scalar4 vel_com;
+            if (mass + mass_cell > 0)
+                {
+                vel_com.x = (vel_mass.x * mass + vel_mass_cell.x * mass_cell) / (mass + mass_cell);
+                vel_com.y = (vel_mass.y * mass + vel_mass_cell.y * mass_cell) / (mass + mass_cell);
+                vel_com.z = (vel_mass.z * mass + vel_mass_cell.z * mass_cell) / (mass + mass_cell);
+                vel_com.w = mass + mass_cell;
+
+                vel_mass.x += (vel_mass.x - vel_com.x) * factor;
+                vel_mass.y += (vel_mass.y - vel_com.y) * factor;
+                vel_mass.z += (vel_mass.z - vel_com.z) * factor;
+                h_velocity.data[idx] = vel_mass;
+
+                // rescale the other particles in the cell
+                const unsigned int np = h_cell_size.data[bin_idx];
+                const unsigned int N_mpcd = m_mpcd_pdata->getN() + m_mpcd_pdata->getNVirtual();
+                for (unsigned int offset = 0; offset < np; ++offset)
+                    {
+                    // Load particle data
+                    // NOTE: unsure if bin_idx is correct here
+                    const unsigned int cur_p = h_cell_list.data[cli(offset, bin_idx)];
```

Comment:
I think is right, so long as you double checked you have the indexes being passed to `cli` in the same order as they are used on the CPU.